keyword.go

```diff
@@ -0,0 +1,189 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	jptr "github.com/qri-io/jsonpointer"
+)
+
+var notSupported = map[string]bool{
+	// "$schema": true,
+	// "$id": true,
+	"$anchor": true,
+	"$recursiveAnchor": true,
+	"$defs": true,
+	"definitions": true,
+	"$ref": true,
+	"$recursiveRef": true,
+	// "title": true,
+	// "description": true,
+	// "default": true,
+	// "examples": true,
+	// "readOnly": true,
+	// "writeOnly": true,
+	// "$comment": true,
+	"$vocabulary": true,
+
+	// array keywords
+	// "items": true,
+	// "additionalItems": true,
+	// "maxItems": true,
+	// "minItems": true,
+	// "uniqueItems": true,
+	// "contains": true,
+	// "maxContains": true,
+	// "minContains": true,
+	"unevaluatedItems": true,
+
+	// object keywords
+	"unevaluatedProperties": true,
+
+	// conditional keywords
+	"if": true,
+	"then": true,
+	"else": true,
+
+	// other
+	"contentEncoding": true,
+	"contentMediaType": true,
+	"contentSchema": true,
+	"dependentRequired": true,
+	"dependentSchemas": true,
+	"deprecated": true,
+}
+
+var KeywordRegistry = map[string]KeyMaker{}
+var KeywordOrder = map[string]int{}
+
+func IsKeyword(prop string) bool {
+	_, ok := KeywordRegistry[prop]
+	return ok
+}
+
+func GetKeyword(prop string) Keyword {
+	if !IsKeyword(prop) {
+		return NewVoid()
+	}
+	return KeywordRegistry[prop]()
+}
+
+func GetKeywordOrder(prop string) int {
+	if order, ok := KeywordOrder[prop]; ok {
+		return order
+	}
+	return 1
+}
+
+func SetKeywordOrder(prop string, order int) {
+	KeywordOrder[prop] = order
+}
+
+func IsNotSupportedKeyword(prop string) bool {
+	_, ok := notSupported[prop]
+	return ok
+}
+
+func RegisterKeyword(prop string, maker KeyMaker) {
+	KeywordRegistry[prop] = maker
+}
+
+// MaxValueErrStringLen sets how long a value can be before it's length is truncated
+// when printing error strings
+// a special value of -1 disables output trimming
+var MaxKeywordErrStringLen = 20
+
+// Validator is an interface for anything that can validate.
+// JSON-Schema keywords are all examples of validators
+type Keyword interface {
+	// Validate checks decoded JSON data and writes
+	// validation errors (if any) to an outparam slice of ValErrors
+	// propPath indicates the position of data in the json tree
+	Validate(propPath string, data interface{}, errs *[]KeyError)
+	ValidateFromContext(schCtx *SchemaContext, errs *[]KeyError)
+
+	Register(uri string, registry *SchemaRegistry)
+	Resolve(pointer jptr.Pointer, uri string) *Schema
```

Comment:
if either of these need to make HTTP calls, they should accept a `context.Context` argument, and use that context when performing the request: https://golang.org/pkg/net/http/#NewRequestWithContext

Reply:
Good point. Network and local calls will be intermingled. I'll either pass the ctx along in the function call or keep a reference in the schema context.

Reply:
Definitely go with the former. As per the [context package docs](https://golang.org/pkg/context/):

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.

Reason for this: multiple different contexts may touch the same parsed schema. Retaining a context forces you to throw away the struct to use a new context

Reply:
I took an inbetween approach. Since the call tree is highly nested and users generally don't reach that far down to invoke the actions, I've used the schema context we're passing around as the carrier and only pass on the context to the functions explicitly requiring it.

This should work fine as the same context is carried throughout the call stack.